lib/lfrfid/lfrfid_worker_modes.c

```diff
@@ -0,0 +1,514 @@
+#include <furi.h>
+#include <furi_hal.h>
+#include "lfrfid_worker_i.h"
+#include "tools/t5577.h"
+#include <stream_buffer.h>
+#include <toolbox/pulse_protocols/pulse_glue.h>
+
+#define TAG "LFRFIDWorker"
+
+/**
+ * if READ_DEBUG_GPIO is defined:
+ *     gpio_ext_pa7 will repeat signal coming from the comparator
+ *     gpio_ext_pa6 will show load on the decoder
+ */
+#define LFRFID_WORKER_READ_DEBUG_GPIO 1
+
+#ifdef LFRFID_WORKER_READ_DEBUG_GPIO
+#define LFRFID_WORKER_READ_DEBUG_GPIO_VALUE &gpio_ext_pa7
+#define LFRFID_WORKER_READ_DEBUG_GPIO_LOAD &gpio_ext_pa6
+#endif
+
+#define LFRFID_WORKER_READ_DROP_TIME 500
+#define LFRFID_WORKER_STABILIZE_TIME 1000
+#define LFRFID_WORKER_READ_SWITCH_TIME 2000
+
+#define LFRFID_WORKER_WRITE_VERIFY_TIME 2000
+#define LFRFID_WORKER_WRITE_DROP_TIME 500
+
+#define LFRFID_WORKER_WRITE_TOO_LONG_TIME 10000
+#define LFRFID_WORKER_WRITE_MAX_UNSUCCESSFUL_READS 5
+
+#define LFRFID_WORKER_READ_STREAM_SIZE 4096
+#define LFRFID_WORKER_EMULATE_BUFFER_SIZE 1024
+
+/**************************************************************************************************/
+/********************************************** READ **********************************************/
+/**************************************************************************************************/
+
+static void lfrfid_worker_read_capture(bool level, uint32_t duration, void* context) {
+#ifdef LFRFID_WORKER_READ_DEBUG_GPIO
+    furi_hal_gpio_write(LFRFID_WORKER_READ_DEBUG_GPIO_VALUE, level);
+#endif
+
+    StreamBufferHandle_t stream = context;
+    LevelDuration level_duration = level_duration_make(level, duration);
+    BaseType_t xHigherPriorityTaskWoken = pdFALSE;
+
+    xStreamBufferSendFromISR(
+        stream, &level_duration, sizeof(LevelDuration), &xHigherPriorityTaskWoken);
+    portYIELD_FROM_ISR(xHigherPriorityTaskWoken);
+}
+
+typedef enum {
+    LFRFIDWorkerReadOK,
+    LFRFIDWorkerReadExit,
+    LFRFIDWorkerReadTimeout,
+} LFRFIDWorkerReadState;
+
+static LFRFIDWorkerReadState lfrfid_worker_read_internal(
+    LFRFIDWorker* worker,
+    LFRFIDFeature feature,
+    uint32_t timeout,
+    ProtocolId* result_protocol) {
+    LFRFIDWorkerReadState state;
+    furi_hal_rfid_pins_read();
+
+    if(feature & LFRFIDFeatureASK) {
+        furi_hal_rfid_tim_read(125000, 0.5);
+        FURI_LOG_D(TAG, "Start ASK");
+    } else {
+        furi_hal_rfid_tim_read(62500, 0.25);
```

Comment:
magic

Reply:
Literally frequency and duty cycle, as furi_hal_rfid_tim_read arguments says. I don't see the point in naming it.